### PR TITLE
Bits per sample

### PR DESCRIPTION
**Pull request by MathemanFlo**
Implements https://github.com/libvips/libvips/discussions/3321

- I added a `VIPS_META_BITS_PER_SAMPLE` flag which contains the bits per pixel of each channel. Hope you are okay with the name. Alternative would be `VIPS_META_BITS_PER_CHANNEL` or `VIPS_META_BITS_PER_BITDEPTH`. Avoided the latter because it is unclear whether it is per channel or the total color depth.
- Support for tiffs via `TIFFTAG_BITSPERSAMPLE`. According to [libtiff reference](https://www.awaresystems.be/imaging/tiff/tifftags/bitspersample.html) a tiff file could contain different bits per sample for each channel but libtiff does not support that, so it is okay to only read the tag once. 
- Support for libpng and libspng. [libpng reference](https://refspecs.linuxfoundation.org/LSB_5.0.0/LSB-TrialUse/LSB-TrialUse/libpng15-png-get-ihdr.html)

I noticed that the code for png and gif contains `"palette-bit-depth"` which is very similar to `VIPS_META_BITS_PER_SAMPLE`.
Haven't looked at gif and webp yet, but will asap.

### CODE REVIEW

**Comment by kleisauke** [comment]
> I noticed that the code for png and gif contains `"palette-bit-depth"` which is very similar to `VIPS_META_BITS_PER_SAMPLE`.

Good point, I think the best way to deal with this is just to deprecate `palette-bit-depth` in favor of `bits-per-sample` (`VIPS_META_BITS_PER_SAMPLE`). Deprecated header fields are tracked here: https://github.com/libvips/libvips/blob/76f2178a31d62ebc84e30a0d0a04216271bd7e34/libvips/iofuncs/header.c#L1454-L1460

File: libvips/foreign/nsgifload.c
Comment (kleisauke) [inline libvips/foreign/nsgifload.c:335]: I think we still need to set the `palette-bit-depth` field for compatibility.

File: libvips/foreign/spngload.c
Comment (kleisauke) [inline libvips/foreign/spngload.c:276]: ditto

File: libvips/foreign/vipspng.c
Comment (kleisauke) [inline libvips/foreign/vipspng.c:599]: ditto

**Review by kleisauke — COMMENTED** [review]
Thanks for working on this! I left a comment about compatibility regarding the `palette-bit-depth` field (for the same reason we still set need to set the `gif-loop` and `gif-delay` fields, unfortunately).

> * As far as I understand WEBP can not have different values for bits per sample as assumed in [Accessing tiff's bits per sample #3321 (comment)](https://github.com/libvips/libvips/discussions/3321#discussioncomment-4849971) . According to my research it is always 8 bit per channel.

Maybe we should always set it to 8 for WebP images?

> * We potentially could support HEIC/AVIF. Look for `"heif-bitdepth"` and `bits_per_pixel` in heifload.c
> * We could also support JPEG2000

I think this would be also useful to have :+1:. JXL images could support this too (via `jxl->info.bits_per_sample`).

File: libvips/foreign/nsgifload.c
Comment (MathemanFlo) [reply libvips/foreign/nsgifload.c:337]: I fear I miunderstood how deprecation works. I added the deleted code lines again.

**Comment by kleisauke** [comment]
> Added support for jpeg xl, heif/avif, jpeg 2000

Great!

> This is indeed possible but then the question arises why do it for WebP and not for all other file formats like jpeg (which also only supports 8 bits per pixel per channel)

I see. It probably makes more sense to avoid setting this for images that do not support it.

**Review by kleisauke — APPROVED** [review]
LGTM, I've updated the PR to support this within the *magick loader and deprecated the `heif-bitdepth` meta field while we're here. I also covered this with a few unit tests.

I'll leave this here for a few days for any further feedback from other maintainers before merging. Thank you!

**Review by lovell — APPROVED** [review]
LGTM, thanks both! There will be (optional) support for 12-bit JPEG in the forthcoming libjpeg-turbo v3, but that's very much a future possible enhancement.

**Pull request merged**
